### PR TITLE
Public memory, part 1

### PR DESCRIPTION
**Pull request by dlubarov**
Sorry for the large PR; a lot of it is just ripping out the existing public input related code. The main new addition is `PublicValues`, which contains trie roots (before and after) and block metadata. These values could be thought of as public inputs which live in memory, but I wanted to avoid "input" to avoid confusion with `GenerationInputs`.

This doesn't yet include the change to the `Z(last)` comparison, where the verifier itself will (in a sense) read the public values, by computing the (reduced) CTL rows for those special read operations and factoring them into the `Z(last)` comparison. I'll add that later.

### CODE REVIEW

File: evm/src/proof.rs [hunk 1]
```diff
@@ -42,7 +44,56 @@ pub(crate) struct AllProofChallenges<F: RichField + Extendable<D>, const D: usiz
...
+
+#[derive(Debug, Clone, Default)]
+pub struct TrieRoots {
+    pub state_root: U256,
+    pub transactions_root: U256,
+    pub receipts_root: U256,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct BlockMetadata {
```
Comment (wborgeaud) [inline evm/src/proof.rs:67]: Do we also need a `block_basefee` here to support the `BASEFEE` opcode?

File: evm/src/util.rs [hunk 2]
```diff
@@ -40,3 +41,29 @@ pub fn trace_rows_to_poly_values<F: Field, const COLUMNS: usize>(
         .map(|column| PolynomialValues::new(column))
         .collect()
 }
+
+/// Returns the 32-bit little-endian limbs of a `U256`.
+pub(crate) fn u256_limbs<F: Field>(u256: U256) -> [F; 8] {
+    u256.0
+        .into_iter()
+        .flat_map(|limb_64| {
+            let lo = (limb_64 & 0xFFFFFFFF) as u32;
```
Comment (wborgeaud) [inline evm/src/util.rs:50]: nit: Isn't that the same as `limb_64 as u32`? Or do you think it's clearer that way?

**Review by wborgeaud — APPROVED** [review]
Looks good to me!

File: evm/src/util.rs (hunk 2 above)
Comment (dlubarov) [reply evm/src/util.rs:50]: Ah, somehow I thought it would fail in debug mode, but looks like that's not the case.

File: evm/src/proof.rs (hunk 1 above)
Comment (dlubarov) [reply evm/src/proof.rs:67]: I missed that, thanks!

**Pull request merged**